[만들면서 배우는 DI 2,3단계] 밀러 미션 제출합니다

oungsi2000 opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    -  [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
- 안녕하세요 2단계도 잘 부탁드립니다 ~
di를 별도의 모듈로 분리하다 보니 file change가 상당하네요..
jitpack에 배포해서 라이브러리 사용 가능합니다

root의 settings.gradle에서
```kotlin
dependencyResolutionManagement {
		repositoriesMode.set(RepositoriesMode.FAIL_ON_PROJECT_REPOS)
		repositories {
			mavenCentral()
			maven { url 'https://jitpack.io' }
		}
	}
```

다음과 같은 코드 추가 이후 의존성 추가 가능합니다


```kotlin
implementation("com.github.oungsi2000:compact-android-di:beta-1.0.0")
```

현재는 동시성 문제를 고려하지 않고 있습니다
추후 동시성 문제도 고려해보도록 하겠습니다


## 코드 리뷰 커뮤니케이션

 <!--
 리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
 특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
 -->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->

![]()

oungsi2000 committed:
refactor: CartViewModel이 Repository 인터페이스를 의존하도록 수정

- CartViewModel이 구체 클래스(`DefaultCartRepository`) 대신 인터페이스(`CartRepository`)에 의존하도록 변경했습니다.
- 불필요해진 `AppContainerDelegate`를 삭제했습니다.

oungsi2000 committed:
test: 테스트용 Product 객체를 Fixture로 분리

- 테스트 코드에서 반복적으로 사용되는 `Product` 객체를 `PRODUCT1`이라는 Fixture로 분리하여 재사용성을 높였습니다.
- `CartRepositoryTest`, `MainViewModelTest` 등에서 하드코딩된 `Product` 생성을 Fixture로 대체했습니다.
- `Robolectric`과 `InstantTaskExecutorRule` 등 불필요한 테스트 의존성을 제거하여 테스트 환경을 개선했습니다.

oungsi2000 committed:
feat: Room을 사용하여 장바구니 데이터 영속성 구현

- Room 데이터베이스를 설정하고 CartProductDao를 `DefaultCartRepository`에 주입했습니다.
- CartRepository의 모든 메서드를 suspend 함수로 변경하여 비동기 데이터 처리를 적용했습니다.
- ViewModel에서 코루틴 스코프(viewModelScope)를 사용하여 CartRepository의 suspend 함수를 호출하도록 수정했습니다.
- Product 모델에 생성 시간(`createdAt`)을 추가하고, 장바구니 화면에 상품이 담긴 시간을 표시하도록 구현했습니다.

oungsi2000 committed:
feat: 상품의 position과 deleteCartProduct의 id가 동일한 값임을 보장

- 상품을 식별하기 위해 `Product` 모델에 `id` 필드를 추가했습니다.
- 장바구니에서 상품을 삭제할 때 기존의 `position` 기반이 아닌, `id`를 사용하도록 로직을 수정했습니다.
- `CartProductAdapter`와 `CartProductViewHolder`에서 `id`를 이용해 삭제를 처리하도록 변경했습니다.
- `DefaultProductRepository`의 더미 데이터와 `ProductMapper`에 `id`를 반영했습니다.

oungsi2000 committed:
refactor: DI 컨테이너를 초기화하는 방식 DSL로 변환

- DI 컨테이너를 `DiApplication`이라는 추상 클래스로 분리했습니다.
- `DependencyFactory` 생성 방법을 `factory` 함수로 단순화했습니다.
- `vararg`를 사용하여 `AppContainerStore`의 생성자 파라미터를 유연하게 변경했습니다.
- `MainApplication`에서 새로운 DI 초기화 방식을 적용했습니다.
- ViewModel 자동 주입 로직을 `DiApplication`을 사용하도록 수정했습니다.

oungsi2000 committed:
refactor: step3, Qualifier 구현

oungsi2000 committed:
refactor: step2, 필드 주입 구현

oungsi2000 committed:
feat: ViewModel 자동 생성을 위한 @AutoViewModel 어노테이션 추가

- ViewModel 자동 생성을 제어하기 위해 `@AutoViewModel` 어노테이션을 추가했습니다.
- `@AutoViewModel`이 붙은 ViewModel은 인스턴스화 될 때 캐시에 저장되지 않도록 `AppContainerStore`의 로직을 수정했습니다.
- `MainViewModel`과 `CartViewModel`에 `@AutoViewModel` 어노테이션을 적용했습니다.

oungsi2000 committed:
feat: DI 관련 코드 CompactAndroidDI 라이브러리 모듈로 분리

기존 app 모듈에 있던 DI(Dependency Injection) 관련 클래스들을 `CompactAndroidDI`라는 별도의 안드로이드 라이브러리 모듈로 이전했습니다.

- `CompactAndroidDI` 라이브러리 모듈 및 관련 설정 파일을 추가했습니다.
- DI 관련 클래스들(`AppContainerStore`, `DiApplication`, `AutoViewModel` 등)을 새 모듈로 이동하고 패키지명을 `com.daedan.compactAndroidDi`로 변경했습니다.
- app 모듈에서는 `CompactAndroidDI` 모듈을 implementation으로 참조하도록 수정했습니다.
- DI 기능 이동에 따라 app 모듈 내 관련 클래스들의 import 경로를 업데이트했습니다.

oungsi2000 committed:
test: Di 테스트와 프로덕션 테스트 분리 및 테스트 통과되게 변경

- `MainViewModelTest`에서 `runTest`와 `TestDispatcher`를 사용하여 코루틴 테스트 환경을 구성했습니다.
- `CartRepositoryTest`가 `FakeCartProductDao`를 사용하도록 수정하고, 코루틴 테스트를 위해 `runTest`를 적용했습니다.
- 테스트의 독립성을 보장하기 위해 `FakeCartProductDao`를 구현하여 메모리 내에서 동작하도록 했습니다.
- Repository의 suspend 함수 변경에 따라 `FakeCartRepository`의 메서드들을 suspend 함수로 수정했습니다.
- 불필요한 `ExampleUnitTest.kt` 파일을 삭제했습니다.

oungsi2000 committed:
refactor: 순환 참조 감지를 위해 inProgress를 멤버 변수로 변경

- 인스턴스화 과정에서 순환 참조를 감지하는 `inProgress` Set을 `instantiate` 함수의 지역 변수에서 `AppContainerStore`의 멤버 변수로 변경했습니다.
- 스레드 안전성을 보장하기 위해 `Collections.synchronizedSet`으로 `inProgress`를 래핑했습니다.
- `inProgress`가 멤버 변수가 됨에 따라, `instantiate`, `reflect`, `injectField` 등 관련 함수들의 파라미터에서 `inProgress`를 제거했습니다.

oungsi2000 committed:
test: CompactAndroidDI 테스트 수정

`AppContainerStore`가 `DependencyFactory` 대신 `DependencyModule`을 통해 의존성을 등록하도록 변경했습니다. 이를 통해 모듈 단위로 의존성을 그룹화하고 관리할 수 있습니다.

- `AppContainerStore.registerFactory()`가 `DependencyModule`을 받도록 수정
- `DiApplication.register()` 로직 간소화
- 관련 테스트 코드를 모듈 기반으로 리팩토링 및 테스트 케이스 보강

oungsi2000 committed:
chore: KtlintFormat

oungsi2000 committed:
refactor: 테스트와 앱에서 공통으로 사용하는 module 함수 분리

테스트 코드(`TestUtil.kt`)에 있던 `module` 함수를 `DependencyModule.kt`로 이동시켜, 테스트와 프로덕션 코드에서 모두 사용할 수 있도록 변경했습니다.

oungsi2000 committed:
chore: JitPack 배포를 위한 설정 추가

- JitPack을 통해 라이브러리를 배포할 수 있도록 `maven-publish` 플러그인과 관련 설정을 추가했습니다.
- `settings.gradle.kts`에 jitpack.io 저장소를 추가했습니다.
- app 모듈에서 로컬 프로젝트 의존성을 제거하고 JitPack을 통해 라이브러리를 참조하도록 수정했습니다.

oungsi2000 committed:
docs: 해결한 문제 목록 README에 추가

oungsi2000 committed:
feat: Maven 배포 정보 수정

- `artifactId`를 `CompactAndroidDI`에서 `compact-android-di`로 변경했습니다.
- `version`을 `1.0.0`에서 `beta-1.0.0`으로 변경했습니다.

hwannow reviewed (changes requested):
안녕하세요 밀러 🌾
2,3 단계도 고생 많으셨습니다 !

이번 미션 개인적으로 아주 어려웠는데 아주 멋지게 해 주셨군요 ...

제가 궁금한 부분들 위주로 리뷰 남겨 놨으니 확인해 주시면 감사하겠습니다 🙇‍♀️

리팩터링도 파이팅입니다 ~ 💪

oungsi2000 committed:
refactor: 모듈 이름을 CompactAndroidDI에서 di로 변경 및 jitpack 배포 설정 추가

- CompactAndroidDI 모듈의 이름을 `di`로 변경했습니다.
- 이에 따라 settings.gradle.kts와 app/build.gradle.kts의 모듈 참조 경로를 수정했습니다.

oungsi2000 committed:
refactor: `@Component` 어노테이션을 이용한 Qualifier 기능 추가

- 의존성 주입 시 특정 구현체를 구분하기 위해 `@Component` 어노테이션을 활용하는 방식을 도입했습니다.
- `@Inject` 어노테이션과 이름 기반의 `Qualifier`를 대체하여 타입 안정성을 높였습니다.
- 기존의 `Qualifier` 데이터 클래스를 `Qualifier` 인터페이스와 여러 구현체(`AnnotationQualifier`, `NamedQualifier`, `TypeQualifier`)로 리팩토링했습니다.
- `DIHelper` 유틸리티를 추가하여 어노테이션 기반으로 `Qualifier`를 결정하는 로직을 구현했습니다.
- `CartRepository`의 의존성 주입 방식을 새로운 `@Component` 어노테이션 방식(`@RoomDBCartRepository`)으로 변경했습니다.

oungsi2000 committed:
feat: Inject 어노테이션의 name 파라미터에 기본값 추가

oungsi2000 committed:
refactor: AppContainerStore의 reflect 메서드 제거

- 기존의 `KClass`와 `name: String` 조합으로 의존성을 식별하던 방식에서 `Qualifier` 인터페이스를 사용하도록 변경했습니다.
- `TypeQualifier`, `NamedQualifier`, `AnnotationQualifier` 등 다양한 종류의 `Qualifier`를 도입하여 의존성을 더 명확하게 구분할 수 있도록 개선했습니다.
- `factory`와 `get` 함수가 `Qualifier`를 직접 받도록 수정하고, 어노테이션 기반 주입을 지원하기 위해 `injectField` 로직을 수정했습니다.
- 이 변경에 따라 `CartViewModel`에서 `@Inject("name")` 대신 커스텀 어노테이션인 `@RoomDBCartRepository`를 사용하도록 수정했습니다.

oungsi2000 committed:
feat: AutoViewModel 어노테이션 제거

ViewModel을 자동으로 주입하던 `@AutoViewModel` 어노테이션과 관련 로직을 제거했습니다.

oungsi2000 committed:
refactor: DI 모듈 분리

DI 로직을 `DataModule`, `RepositoryModule`, `ViewModelModule`로 분리하여 코드를 개선했습니다.

oungsi2000 committed:
refactor: AppContainerStore 순환 참조시 stackoverflow 대신 에러 발생하도록 수정

- getInstance()에서 순환 참조가 발생했을 때 stackoverflow가 발생하는 대신, `inProgress`에서 qualifier를 제거하도록 `finally` 블록을 사용해 수정했습니다.
- 이를 통해 순환 참조 에러 메시지가 정상적으로 표시됩니다.

oungsi2000 committed:
feat: ViewModel 및 Singleton 스코프 추가

- `viewModel`과 `single` 스코프를 추가하여 의존성 주입 시 객체 생성 규칙(스코프)을 지정할 수 있도록 개선했습니다.
- 기존의 `factory` 함수를 `viewModel`로 변경하여 ViewModel 의존성 주입을 명시적으로 표현했습니다.
- `single` 스코프를 도입하고, `RoomDatabase`, `DAO`, `Repository` 등 싱글턴으로 관리되어야 하는 객체들에 적용했습니다.
- 스코프 구분을 위해 `CreateRule` enum (`SINGLETON`, `FACTORY`)을 추가하고 `DependencyFactory`와 `AppContainerStore`에 관련 로직을 구현했습니다.

oungsi2000 committed:
feat: 동일 Qualifier에 대한 중복 등록 방지 로직 추가

- `registerFactory` 메서드 내에서 동일한 Qualifier를 가진 팩토리가 중복으로 등록되는 것을 방지하는 검증 로직을 추가했습니다.
- 중복된 Qualifier가 발견될 경우, `require` 문을 통해 예외를 발생시키고 충돌하는 키 목록을 메시지로 알려줍니다.

oungsi2000 committed:
refactor: private으로 선언된 필드에도 의존성 주입되도록 수정

- MainViewModel의 Repository 필드를 private으로 변경했습니다.
- 의존성 주입 시 private 필드에 접근할 수 있도록 `isAccessible = true`로 설정했습니다.

oungsi2000 committed:
refactor: DI 라이브러리 패키지명 변경

DI 라이브러리의 패키지명을 `com.daedan.compactAndroidDi`에서 `com.daedan.di`로 변경하고, 이에 맞춰 `build.gradle.kts`의 `namespace`와 관련 import 경로들을 수정했습니다.

oungsi2000 committed:
test: MainViewModel 테스트에 DI 컨테이너 적용

- `MainViewModelTest`에서 수동으로 의존성을 주입하던 방식을 DI 컨테이너(`AppContainerStore`)를 사용하도록 변경했습니다.
- 테스트용 `FakeProductRepository`와 `FakeCartRepository`를 DI 모듈에 등록하여 주입받도록 수정했습니다.
- `MainViewModel` 또한 DI 컨테이너를 통해 생성하도록 변경했습니다.

oungsi2000 committed:
feat: NamedQualifier의 name을 non-null로 변경

`NamedQualifier`가 항상 이름을 갖도록 name 프로퍼티의 타입을 `String?`에서 `String`으로 변경했습니다.

oungsi2000 committed:
refactor: 테스트 코드 패키지 구조 변경

- data, ui 레이어에 맞게 테스트 코드의 패키지 구조를 변경했습니다.
- 이에 따라 관련 테스트 클래스의 import 경로를 수정했습니다.

oungsi2000 committed:
test: AppContainerStore에 대한 테스트 코드 추가 및 리팩토링

- `factory`를 `single`, `viewModel`로 분리하고 기능에 맞게 테스트를 수정했습니다.
- 이름 기반(named) 의존성 주입 테스트를 추가했습니다.
- 어노테이션 기반(@Component) 의존성 주입 테스트를 추가했습니다.
- `Qualifier`를 이름, 어노테이션, 타입으로 분리하고 테스트에 적용했습니다.

oungsi2000 committed:
feat: AnnotationQualifier에 Component 어노테이션 검증 로직 추가

oungsi2000 committed:
test: 필드 및 생성자 주입 테스트 코드 추가

- 필드 주입과 생성자 주입을 테스트하기 위한 Fixture 클래스들을 추가했습니다.
- 다양한 주입 시나리오(타입, 이름, 어노테이션 기반)에 대한 테스트 케이스를 작성했습니다.
- 필드 주입, 생성자 주입, 그리고 두 방식이 혼합된 경우에 대한 인스턴스화 테스트를 추가했습니다.

oungsi2000 committed:
refactor: di 모듈을 외부 라이브러리로 분리

- 기존에 프로젝트 내 모듈로 존재하던 di 모듈을 제거하고, 외부 라이브러리 `com.github.oungsi2000:android-di`를 사용하도록 변경했습니다.
- `libs.versions.toml`에 해당 라이브러리를 추가하고, app 모듈의 `build.gradle.kts`에서 의존성을 수정했습니다.

oungsi2000 commented on CompactAndroidDI/src/main/java/com/daedan/compactAndroidDi/AppContainerStore.kt:
```diff
@@ -0,0 +1,114 @@
+        save(qualifier, instance)
+        inProgress.remove(qualifier)
+        return instance
+    }
```
[refactor: AppContainerStore 순환 참조시 stackoverflow 대신 에러 발생하도록 수정](https://github.com/woowacourse/android-di/pull/199/commits/736cd779190dcd8975b4f36969869a295779824b)

oungsi2000 commented on CompactAndroidDI/src/main/java/com/daedan/compactAndroidDi/Qualifier.kt:
```diff
@@ -0,0 +1,8 @@
+data class Qualifier(
+    val type: KClass<*>,
+    val name: String? = null,
+)
```
구현했을 당시에는 Annotation은 메타데이터이기 떄문에 인자로 넣을 수 없다고 알고 있어서 리플렉션에 기반한 런타임 DI에는 조금 구현이 어렵지 않나 싶었습니다. 

다만 다시 알아보니 Annotation이라는 클래스로 인자를 받을 수 있더라구요!

반영했습니다 ~!
[refactor: @Component 어노테이션을 이용한 Qualifier 기능 추가](https://github.com/woowacourse/android-di/pull/199/commits/691b22a53035562965088f1f2da51f271ee92a28)

oungsi2000 commented on CompactAndroidDI/src/main/java/com/daedan/compactAndroidDi/annotation/AutoViewModel.kt:
```diff
@@ -0,0 +1,4 @@
+package com.daedan.compactAndroidDi.annotation
+
+@Target(AnnotationTarget.CLASS)
+annotation class AutoViewModel
```
사실은...reflect 메서드가 처리할 타입을 명확히 하기 위해서였는데요. 저도 환노의 말씀대로 확장성이 떨어진다고 생각해요

말이 길어질 것 같습니다..

### reflect 메서드의 탄생 배경
처음에는 팩토리에 등록되지 않은 객체라면, AppContainerStore에 요청한 객체의 생성자를 분석하여, 재귀적으로 reflect를 호출하여 객체를 만들고 있었어요. 

제가 많이 참고한 DI 라이브러리가 Koin인데, Koin은 ViewModel을 포함한 모든 의존성을 module에 등록해줘야 했지만, 일일이 모든 의존성을 등록하는 것이 불편하다고 느껴져 이렇게 구현했습니다.

하지만 이렇게 하니 복잡한 의존 관계가 엮여 있을 때,  어떤 것은 팩토리에 등록되있고, 어떤 것은 등록되어있지 않아, 오히려 일관성이 떨어지고, 훨씬 불편하다고 느껴졌어요. 

### 어노테이션으로 reflect 대상 제한
이것을 해결하기 위해 primaryConstructor가 아닌, 어노테이션이 붙은 생성자만 reflect의 대상이 되도록 구현했습니다. 

처음에는 SingleTon, AutoViewModel, Scope 이 세 개의 어노테이션으로 각각, SingleTon은 캐시에 저장, AutoViewModel은 저장 안함, Scope는 추후 4단계에 스코프 설정

이렇게 설계를 했었는데요. 
하지만 저번에 말씀드린 것처럼 사실상 reflect 메서드는 ViewModel생성에만 이용하고 있고, 나머지는 초기 module에 등록하고 있어요. 

또 초기 module에 등록과 어노테이션을 중복으로 등록할 수 있어 이렇게 되면 또 일관성이 떨어지는 문제가 있다고 생각했어요. 

때문에 SingleTon, Scope 어노테이션은 쓰지 않는 어노테이션이 되어 폐기 처리 하도록 하였습니다. 

### 개선 
이렇다 보니 결국 남은건 AutoViewModel 어노테이션인지라. 가독성을 높이려고 하다 보니 환노 말씀대로 확장성이 떨어지게 되었네요. 

지금은 reflect 메서드 제거, AutoViewModel을 제거하고, viewModel 또한 초기 모듈로 무조건 등록하게 바꿨습니다. 

[refactor: AppContainerStore의 reflect 메서드 제거](https://github.com/woowacourse/android-di/pull/199/commits/c07c616b6756c45cdbaca2cc3cd033f1dff8ecef)

[feat: AutoViewModel 어노테이션 제거](https://github.com/woowacourse/android-di/pull/199/commits/d672d9cb382c34a0c74757f7e08acf11b7c1592b)

oungsi2000 committed:
refactor: AppContainerStore에서 사용하지 않는 get 연산자 제거

oungsi2000 commented on CompactAndroidDI/src/main/java/com/daedan/compactAndroidDi/AppContainerStore.kt:
```diff
@@ -0,0 +1,114 @@
+            mutableSetOf(),
+        )
+
+    operator fun get(clazz: KClass<*>): Any? = cache[Qualifier(clazz)]
```
[refactor: AppContainerStore에서 사용하지 않는 get 연산자 제거](https://github.com/woowacourse/android-di/pull/199/commits/792c290fcbc739ab9af12e0950582aab483d3044)

oungsi2000 commented on app/src/main/java/woowacourse/shopping/ui/MainViewModel.kt:
```diff
@@ -3,23 +3,33 @@ package woowacourse.shopping.ui
+
+    @Inject("cartRepository")
+    lateinit var cartRepository: CartRepository
+
```
[refactor: private으로 선언된 필드에도 의존성 주입되도록 수정](https://github.com/woowacourse/android-di/pull/199/commits/11b3f30e4c90a68b985657c5ae47eead70e47f3d)

반영했습니다 ~

oungsi2000 reviewed (commented):
안녕하세요. 
현재 라이브러리 배포 관련하여 패키지를 변경하게 되어 파일 changes가 매우 많습니다. 

app 모듈은 di의 변경사항 (어노테이션 적용 등)을 제외하고는 변경사항 없습니다. 

드디어. jitpack에 라이브러리 배포 완료했습니다
현재 커밋에는 배포판이 적용되어 있으며, 
https://jitpack.io/#oungsi2000/android-di/beta-1.0.0
해당 문서를 보면 제 라이브러리 설치 가능합니다 ~

모듈을 생성할 때 single, viewModel 이렇게 두 가지 dsl로 Dependency 팩토리를 생성할 수 있는데요.
현재는 viewModel dsl이 사실 객체를 매번 생성한다는 역할만 부여받았을 뿐, 
viewModel이 아닌 객체에도 활용할 수 있습니다. 

Koin은 대표적으로 single, factory, viewModel 세 가지 dsl을 제공하는데요.
**single은 싱글톤, factory는 매번 인스턴스 생성, viewModel은 VIewModelFactory 저장**
이렇게 동작한다고 알고 있습니다. 

지금 제가 구현한 viewModel dsl은 사실 Koin에 factory에 대응하고 있습니다. 혼동을 방지하기 위해(이미 많이 혼동을 드린 것 같지만...) 일단 single과 viewModel dsl만 구현했습니다

4단계 때에는 Koin과 동일하게 구현해 보겠읍니다...

## 라이브러리 구현하면서 느낀 점
이번 DI 미션을 구현하면서 느낀 점을 써볼까 하는데요. 
(주저리주저리 글이라 스킵하셔도 무방합니다)

가장 첫 번째로 고민한 점은
**의존성을 수동으로 등록해줄 것인가, 어노테이션을 사용하여 Module을 찾아서 의존성을 등록할 것인가. **

였습니다. 
그래서 Java Reflection 라이브러리 중, 
클래스패스 스캐닝을 쉽게 처리할 수 있는 org.reflections 라이브러리가 존재했습니다. 
이를 런타임에 컴포넌트를 스캔하여 의존성을 등록하도록 고민해 보았는데요. 
문제가 있었습니다. 

안드로이드는 패키징 시, jar이 아닌 dex로 컴파일하여 디렉토리를 이리저리 바꾸게 됩니다. 
만약 난독화까지 적용하게 된다면 꽤나 성능이 나빠질 것 같다는 생각이 들었습니다. 

그렇다고 이것을 KSP를 이용하여 컴파일 타임에 팩토리를 만들자니니...
너무 오버엔지니어링이 아닌가.. 하는 생각도 들었습니다. 

Hilt, Koin과 Sptring의 차이점은 꼭 Module로 의존성을 등록해야 한다는 점입니다. 
Spring은 등록하지 않고 의존성을 등록, 사용할 객체에 ```@Component``` 어노테이션만 붙이거나, 
팩토리 메서드에 ```@Bean```을 붙여 팩토리를 등록하기만 하면 됩니다. 

이런 차이의 원인이 클래스패스 스캐닝과 관련한 문제가 아닐까 조심스레 생각해봅니다. 

하지만 이번 미션을 구현하면서 저는 일일이 Module에 의존성을 지정하는 것이 번거롭다고 느꼈습니다. 
하지만 클래스패스 스캐닝을 쓰기는 어렵고...하니, 생각했던 것이 최소한의 의존성 (예를 들면 ViewModel이나 생성자가 아닌 팩토리 메서드로만 생성해야 하는 객체)만 등록하고, 나머지는 생성자를 재귀적으로 찾아서 객체를 저장하자 ! 라는 생각이었습니다. 

하지만 이렇게 하다 보니 여러 문제를 맞닥뜨렸습니다. (커밋 참조)
결국 모든 의존성을 진입점에 등록하는 것이 제일 최선이라는 생각이 들었습니다. Koin, Hilt가 왜 이렇게 구현을 해 놓았는지 어렴풋이 알게 된 것 같았습니다. 

결국 현재 수업 시간에 리플렉션을 배웠으니, 수업 내용을 크게 벗어나지 않는 선에서 구현할 수 있는 가장 참고할만한 DI 라이브러리는 Koin이라고 생각하여 (런타임, 리플렉션 기반) Koin을 토대로 구현해 보았습니다.

참 어렵네요. 

## +추가) 내가 만든 DI 라이브러리는 서비스 로케이터인가?
서비스 로케이터이지만 제어의 역전을 구현했다고 생각합니다. 

현재는 AppContainerStore에 Map 형태로 인스턴스를 모두 저장(중앙 집중)하지만
등록한 팩토리를 사용자가 아닌 라이브러리가 호출하여 인스턴스를 생성한다는 점에서 Ioc는 달성했다고 생각합니다

그리고 서비스 로케이터 패턴을 사용하지 않으면서 런타임 DI를 구현하는것은 안드로이드 환경에서는 어렵다고 생각합니다. 
이유는 클래스패스 스캐닝을 사용하기 힘들고 이걸 사용하지 않으면 본문에서 말씀드린 문제들이 발생하기 때문에...

hwannow reviewed (approved):
안녕하세요 밀러 🌾

리팩터링을 거치고 나니 한층 더 koin 스러워졌군요
멋진데요 💪

다음 미션을 위해 이번 미션은 머지하도록 하겠습니다 !
4단계도 파이팅입니다 ~
